lib/ide/src/controller/module.rs

```diff
@@ -28,32 +56,161 @@ impl Location {
+// =========================
+
 shared! { Handle
     /// State data of the module controller.
     #[derive(Debug)]
     pub struct Controller {
         /// This module's location.
-        location : Location,
+        location: Location,
+        /// The current module ast, used by synchronizing both module representations.
+        ast: Ast,
+        /// The id map of current ast
+        // TODO: written for test purposes, should be removed once generating id_map from AST will
+        // be implemented.
+        id_map: IdMap,
+        /// The File Manager Client handle.
+        file_manager: fmc::Handle,
+        /// The Parser handle
+        parser: SharedParser,
+        logger: Logger,
     }
 
     impl {
-        /// Create new Module controller for given location.
-        pub fn new(location:Location) -> Self {
-            Controller {location:location}
-        }
-
         /// Obtain clone of location.
         pub fn location(&self) -> Location {
             self.location.clone()
         }
+
+        /// Updates AST after code change.
+        pub fn apply_code_change(&mut self,change:&TextChangedNotification) {
+            let mut code        = self.code();
+            let replaced_range  = change.replaced_chars.clone();
+            let inserted_string = change.inserted_string();
+            let replaced_size   = Size::new(replaced_range.end - replaced_range.start);
+            let replaced_span   = Span::new(Index::new(replaced_range.start),replaced_size);
+
+            code.replace_range(replaced_range,&inserted_string);
+            apply_code_change_to_id_map(&mut self.id_map,&replaced_span,&inserted_string);
+            self.ast = self.parser.parse(code, self.id_map.clone()).unwrap();
+            self.logger.trace(|| format!("Applied change; Ast is now {:?}", self.ast));
+        }
+
+        /// Read module code.
+        pub fn code(&self) -> String {
+            self.ast.repr()
+        }
+
+        /// Check if current module state is synchronized with given code. If it's not, log error,
+        /// and update module state.
+        pub fn check_code_sync(&mut self, code:String) -> FallibleResult<()> {
+            let my_code = self.code();
+            if code != my_code {
+                self.logger.error(|| format!("The module controller ast was not synchronized with
+                    text editor content!\n >>> Module: {:?}\n >>> Editor: {:?}",my_code,code));
+                self.ast    = self.parser.parse(code,default())?;
+                self.id_map = default();
+            }
+            Ok(())
+        }
     }
 }
 
 impl Handle {
-    /// Receives a notification call when file with this module has been
-    /// modified by a third-party tool (like non-IDE text editor).
-    pub async fn file_externally_modified(&self) {
-        // TODO: notify underlying text/graph controllers about the changes
-        todo!()
+    /// Create a module controller for given location.
```

Comment:
I'd suggest that for async functions we should also try to document what the function awaits for. 
(something to give user rough understanding what to expect in terms of remote communication when using this function)